syntax = "proto3";

package yandex.cloud.priv.disk_manager.v1;

import "google/protobuf/timestamp.proto";
import "yandex/cloud/api/operation.proto";
import "yandex/cloud/priv/disk_manager/v1/disk.proto";
import "yandex/cloud/priv/operation/operation.proto";

option go_package = "a.yandex-team.ru/cloud/disk_manager/api/yandex/cloud/priv/disk_manager/v1;disk_manager";

// Snapshots are stored in global AZ (for redundancy), but are tied to a concrete disk in regional AZ.
// Creating initial snapshot is slow. The following snapshots will be created (generally) faster thanks
// to incremental snapshotting. Also, since we checkpoint the disk, it's possible to resume the VM
// right after we create the checkpoint (very quick operation) and continue data copying in the background.
service SnapshotService {
    rpc Create(CreateSnapshotRequest) returns (operation.Operation) {
        option (yandex.cloud.api.operation) = {
            metadata: "CreateSnapshotMetadata"
            response: "CreateSnapshotResponse"
        };
    }

    rpc Delete(DeleteSnapshotRequest) returns (operation.Operation) {
        option (yandex.cloud.api.operation) = {
            metadata: "DeleteSnapshotMetadata"
            response: "DeleteSnapshotResponse"
        };
    }

    rpc RestoreDisk(RestoreDiskFromSnapshotRequest) returns (operation.Operation) {
        option (yandex.cloud.api.operation) = {
            metadata: "RestoreDiskFromSnapshotMetadata"
            response: "google.protobuf.Empty"
        };
    }
}

message CreateSnapshotRequest {
    DiskId src = 1;
    string snapshot_id = 2;
    string folder_id = 3;
    bool incremental = 4;
    string operation_cloud_id = 5;
    string operation_folder_id = 6;
    bool zonal = 7;
}

message CreateSnapshotMetadata {
    double progress = 1;
}

message CreateSnapshotResponse {
    int64 size = 1;
    int64 storage_size = 2;
}

message DeleteSnapshotRequest {
    string snapshot_id = 1;
    string operation_cloud_id = 2;
    string operation_folder_id = 3;
}

message DeleteSnapshotMetadata {
    string snapshot_id = 1;
}

message DeleteSnapshotResponse {
    message ChangedSnapshot {
        string snapshot_id = 1;
        google.protobuf.Timestamp timestamp = 2;
        int64 storage_size = 3;
    }
    repeated ChangedSnapshot changed_snapshots = 1;
}

message RestoreDiskFromSnapshotRequest {
    string snapshot_id = 1;
    DiskId disk_id = 2;
    string operation_cloud_id = 3;
    string operation_folder_id = 4;
}

message RestoreDiskFromSnapshotMetadata {
    double progress = 1;
}
